src/main/scala/workflow/WorkflowUtils.scala

```diff
@@ -0,0 +1,124 @@
+    val fitDeps = new ArrayBuffer[Option[Int]]()
+    val instructionIdToNodeId = scala.collection.mutable.Map.empty[Int, Int]
+    instructionIdToNodeId.put(Pipeline.SOURCE, Pipeline.SOURCE)
+
+    for (instruction <- 0 until instructions.length) {
+      instructions(instruction) match {
+        case est: EstimatorNode => Unit
+        case transformer: TransformerNode => Unit
+        case source: SourceNode => {
+          instructionIdToNodeId.put(instruction, nodes.length)
+          nodes.append(source)
+          dataDeps.append(Seq())
+          fitDeps.append(None)
+        }
+        case TransformerApplyNode(transformer, inputs) => {
+          instructions(transformer) match {
+            case transformerNode: TransformerNode => {
+              instructionIdToNodeId.put(instruction, nodes.length)
+              nodes.append(transformerNode)
+              dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+              fitDeps.append(None)
+            }
+            case estimatorFitNode: EstimatorFitNode => {
+              instructionIdToNodeId.put(instruction, nodes.length)
+              // TODO: Get more reasonable label...
+              nodes.append(new DelegatingTransformerNode("Estimator fit"))
+              dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+              fitDeps.append(Some(instructionIdToNodeId(transformer)))
+            }
+            case _ => throw new RuntimeException("Transformer apply instruction must point at a Transformer")
+          }
+        }
+        case EstimatorFitNode(est, inputs) => {
+          val estimatorNode = instructions(est).asInstanceOf[EstimatorNode]
+          instructionIdToNodeId.put(instruction, nodes.length)
+          nodes.append(estimatorNode)
+          dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+          fitDeps.append(None)
+        }
+      }
+    }
+
+    new ConcretePipeline(nodes.toSeq, dataDeps.toSeq, fitDeps.toSeq, nodes.length - 1)
+  }
+
+  def pipelineToInstructions[A, B](pipeline: Pipeline[A, B]): Seq[Instruction] = {
+    val nodes = pipeline.nodes
+    val dataDeps = pipeline.dataDeps
+    val fitDeps = pipeline.fitDeps
+    val sink = pipeline.sink
+
+    pipelineRecurse(sink, nodes, dataDeps, fitDeps, Map(Pipeline.SOURCE -> Pipeline.SOURCE), Seq())._2
+  }
+
+  def pipelineRecurse(
+    current: Int,
+    nodes: Seq[Node],
+    dataDeps: Seq[Seq[Int]],
+    fitDeps: Seq[Option[Int]],
+    nodeIdToInstructionId: Map[Int, Int],
+    instructions: Seq[Instruction]
+  ): (Map[Int, Int], Seq[Instruction]) = {
+    var curIdMap = nodeIdToInstructionId
+    var curInstructions = instructions
+    System.out.println(current)
+    for (dep <- fitDeps(current)) {
+      if (!curIdMap.contains(dep) && dep != Pipeline.SOURCE) {
+        val (newIdMap, newInstructions) = pipelineRecurse(dep, nodes, dataDeps, fitDeps, curIdMap, curInstructions)
+        curIdMap = newIdMap
+        curInstructions = newInstructions
```

Comment:
hmm... there's got to be a more functional way to write this. something like `(fitDeps ++ dataDeps).foldLeft(instructions)(thing)` - tracking this mutable state is already making nervous.